pkg/health/health.go

```diff
@@ -0,0 +1,160 @@
+	metricsValues, err := CollectMetrics(ctx, provider, query, offset, healthCriteria)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not collect metrics")
+	}
+
+	overallResult := Healthy
+	var results []*CheckResult
+	for i, criteria := range healthCriteria {
+		result := determineResult(criteria.Type, criteria.Threshold, metricsValues[i])
+		results = append(results, result)
+
+		if !result.IsCriteriaMet {
+			overallResult = Unhealthy
+
+			logger := logger.WithFields(logrus.Fields{
+				"metricsType": criteria.Type,
+				"threshold":   criteria.Threshold,
+				"actualValue": result.ActualValue,
+			})
+			// If criteria is a latency, we want to log the percentile as well
+			if criteria.Type == config.LatencyMetricsCheck {
+				logger = logger.WithField("percentile", criteria.Percentile)
+			}
+			logger.Debug("criteria was not met")
+		}
+	}
+
+	return &Diagnosis{
+		OverallResult: overallResult,
+		CheckResults:  results,
+	}, nil
+}
+
+// CollectMetrics returns an array of values collected for each of the specified
+// metrics criteria.
+func CollectMetrics(ctx context.Context, provider metrics.Metrics, query metrics.Query, offset time.Duration, healthCriteria []config.Metric) ([]float64, error) {
+	logger := util.LoggerFromContext(ctx)
+	logger.Debug("start collecting metrics")
+	var values []float64
+	for _, criteria := range healthCriteria {
+		var value float64
+		var err error
+
+		switch criteria.Type {
+		case config.LatencyMetricsCheck:
+			value, err = latency(ctx, provider, query, offset, criteria.Percentile)
+			break
+		case config.ErrorRateMetricsCheck:
+			value, err = errorRatePercent(ctx, provider, query, offset)
+			break
+		default:
+			return nil, errors.Errorf("unimplemented metrics %q", criteria.Type)
+		}
+
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to obtain metrics %q", criteria.Type)
+		}
+		values = append(values, value)
+	}
+
+	return values, nil
+}
+
+// determineResult concludes if metrics criteria was met.
+//
+// The returned value also includes a string with details of why the criteria
+// was met or not.
+func determineResult(metricsType config.MetricsCheck, threshold float64, actualValue float64) *CheckResult {
+	result := &CheckResult{ActualValue: actualValue, Threshold: threshold}
+
+	// As of now, the supported health criteria (latency and error rate) need to
+	// be less than the threshold. So, this is sufficient for now but might need
+	// to change to a switch statement when criteria with a minimum threshold is
+	// added.
+	if actualValue <= threshold {
+		result.IsCriteriaMet = true
+	}
+	return result
+}
+
+// latency returns the latency for the given offset and percentile.
+func latency(ctx context.Context, provider metrics.Metrics, query metrics.Query, offset time.Duration, percentile float64) (float64, error) {
+	alignerReducer, err := metrics.PercentileToAlignReduce(percentile)
+	if err != nil {
+		return 0, errors.Wrap(err, "invalid percentile")
```

Comment:
we get the actual percentile value from the user (e.g. p94) however the API might not support that. so if it's an unsupported value, we return a parse error here.

Reply:
however I agree we should probably notice this "parse error" waaay early on, before even starting to diagnose the service.

Reply:
The config checks are done in `main`, though checks for health criteria has not been added yet (#27). This is done here as a defensive approach as we have done with many other things. I would keep the `float64` here since we would need it once we support any percentile value.